stratum/lib/security/credentials_manager_test.cc

```diff
@@ -0,0 +1,172 @@
+// Copyright 2021-present Open Networking Foundation
+// SPDX-License-Identifier: Apache-2.0
+#include "stratum/lib/security/credentials_manager.h"
+
+#include <fstream>
+#include <memory>
+#include <string>
+
+#include "absl/memory/memory.h"
+#include "absl/strings/str_format.h"
+#include "absl/strings/substitute.h"
+#include "gmock/gmock.h"
+#include "grpcpp/grpcpp.h"
+#include "gtest/gtest.h"
+#include "stratum/glue/net_util/ports.h"
+#include "stratum/glue/status/status_test_util.h"
+#include "stratum/lib/security/test.grpc.pb.h"
+#include "stratum/lib/utils.h"
+
+DECLARE_string(ca_cert);
+DECLARE_string(server_key);
+DECLARE_string(server_cert);
+DECLARE_string(test_tmpdir);
+
+namespace stratum {
+namespace {
+
+class TestServiceImpl final : public ::testing::TestService::Service {
+ public:
+  ::grpc::Status Test(::grpc::ServerContext* /*context*/,
+                      const ::testing::Empty* /*request*/,
+                      ::testing::Empty* /*response*/) override {
+    return ::grpc::Status::OK;
+  }
+};
+
+enum class CertFile {
+  kCaCert = 0,
+  kServerCert = 1,
+  kServerKey = 2,
+};
+
+constexpr char kCertDirPrefix[] = "stratum/lib/security/testdata/certs";
+constexpr char kCaCertFile[] = "ca.crt";
+constexpr char kServerCertFile[] = "stratum.crt";
+constexpr char kServerKeyFile[] = "stratum.key";
+
+std::string GetFilename(CertFile f, int i) {
+  const char* file;
+  switch (f) {
+    case CertFile::kCaCert:
+      file = kCaCertFile;
+      break;
+    case CertFile::kServerCert:
+      file = kServerCertFile;
+      break;
+    case CertFile::kServerKey:
+      file = kServerKeyFile;
+      break;
+  }
+  return absl::StrFormat("%s%d/%s", kCertDirPrefix, i, file);
+}
+
+void CopyFile(std::string from, std::string to) {
+  std::ifstream src(from, std::ios::binary);
+  std::ofstream dst(to, std::ios::binary);
+  ASSERT_TRUE(src.is_open());
+  ASSERT_TRUE(dst.is_open());
+  dst << src.rdbuf();
+}
```

Comment:
```suggestion
status CopyFile(const std::string& from, const std::string& to) {
RETURN_IF_ERROR( ReadFileToString());
RETURN_IF_ERROR(WriteStringToFile());

retutn ok;
}
```

Reply:
iostreams are a big can of worms (encoding, error handling, ...) are best avoided if possible. Luckily, the Stratum codebase already contains two convenience functions which allow to implement this easily. Namely, `ReadFileToString` and `WriteStringToFile`.

For reference, from the [Google C++ guide](https://google.github.io/styleguide/cppguide.html#Streams):

> Use streams only when they are the best tool for the job. This is typically the case when the I/O is ad-hoc, local, **human-readable**, and targeted at other developers rather than end-users. Be consistent with the code around you, and with the codebase as a whole; **if there's an established tool for your problem, use that tool instead**.

Reply:
Thanks for the suggestion. Streams should be gone as of the latest commit